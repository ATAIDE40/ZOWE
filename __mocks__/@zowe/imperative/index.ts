/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

/**
 * Mocks all imperative classes specified under __mocks__/__imperative__
 * @returns mocked version of the @zowe/imperative module
 */
const __imperative__get = function () {
  const fs = require.requireActual("fs");
  const imperativePath = require("path").resolve(__dirname, "..", "..", "__imperative__");
  const extraClasses: any = {};
  fs.readdirSync(imperativePath).forEach((file: string) => {
    extraClasses[file.split(".")[0]] = require(`${imperativePath}/${file}`);
  });
  return {
    ...require.requireActual("@zowe/imperative"),
    // Note: Comment out the line below if you do not want to mock
    //       the classes in __mocks__/__imperative__/ for local testing purposes
    ...extraClasses
    // Note: You can still use Imperative.console.info("debug") in your test
  };
}

module.exports = __imperative__get();
