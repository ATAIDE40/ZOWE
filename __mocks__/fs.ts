/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

const __mocks__path = require("path");

let __mocks__fs = jest.genMockFromModule("fs") as any;
const __mocks__oldReadFileSync = require.requireActual("fs").readFileSync;

// This is a custom function that our tests can use during setup to specify
// what the files on the "mock" filesystem should look like when any of the
// `fs` APIs are used.
let __mocks__mockFiles = Object.create(null);
function __mocks__fs____setMockFiles(newMockFiles: { [key: string]: string }) {
    __mocks__mockFiles = Object.create(null);
    for (const file in newMockFiles) {
        const dir = __mocks__path.dirname(file);
        if (!__mocks__mockFiles[dir]) {
            __mocks__mockFiles[dir] = {};
        }
        __mocks__mockFiles[dir][__mocks__path.basename(file)] = newMockFiles[file];
    }
}

// A custom version of `readdirSync` that reads from the special mocked out
// file list set via __setMockFiles
function __mocks__fs__readdirSync(filePath: string) {
    return __mocks__mockFiles[filePath] || {};
}

// A custom version of `existsSync` that reads from the special mocked out
// file list set via __setMockFiles
function __mocks__fs__existsSync(filePath: string) {
    const fileContents = __mocks__fs__readdirSync(__mocks__path.dirname(filePath))[__mocks__path.basename(filePath)];
    if (typeof fileContents === "undefined") {
        return false;
    }
    return true;
}

// A custom version of `readFileSync` that reads from the special mocked out
// file list set via __setMockFiles
function __mocks__fs__readFileSync(filePath: string, encoding?: string) {
    // Don't mock if yargs is trying to load a locale json file
    if (filePath.match(/node_modules.yargs/)) {
        return __mocks__oldReadFileSync(filePath, encoding);
    }

    if (!__mocks__fs__existsSync(filePath)) {
        throw new Error("File not found");
    }
    return __mocks__fs__readdirSync(__mocks__path.dirname(filePath))[__mocks__path.basename(filePath)];
}

// A custom version of `lstatSync` that reads from the special mocked out
// file list set via __setMockFiles
function __mocks__fs__lstatSync(filePath: string) {
    return {
        isFile: () => false
    };
}

__mocks__fs.__setMockFiles = __mocks__fs____setMockFiles;
__mocks__fs.existsSync = __mocks__fs__existsSync;
__mocks__fs.readFileSync = __mocks__fs__readFileSync;
__mocks__fs.readdirSync = __mocks__fs__readdirSync;
__mocks__fs.lstatSync = __mocks__fs__lstatSync;

module.exports = __mocks__fs;
