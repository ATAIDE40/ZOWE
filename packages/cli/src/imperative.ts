/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as path from "path";

import { IImperativeConfig } from "@zowe/imperative";
import { Constants } from "./Constants";
import { ZosFilesOptions } from "./zosfiles/ZosFiles.options";

import { ZosmfSession } from "@zowe/zosmf-for-zowe-sdk";
import { ZosTsoProfile } from "@zowe/zos-tso-for-zowe-sdk";
import { ZosUssProfile } from "@zowe/zos-uss-for-zowe-sdk";

const config: IImperativeConfig = {
    productDisplayName: Constants.DISPLAY_NAME,
    commandModuleGlobs: [
        "daemon/*.definition!(.d).*s",
        "provisioning/*.definition!(.d).*s",
        "workflows/*.definition!(.d).*s",
        "zosconsole/*.definition!(.d).*s",
        "zosfiles/*.definition!(.d).*s",
        "zosjobs/*.definition!(.d).*s",
        "zosmf/*.definition!(.d).*s",
        "zostso/*.definition!(.d).*s",
        "zosuss/*.definition!(.d).*s",
        "zoslogs/*.definition!(.d).*s"
    ],
    rootCommandDescription: Constants.DESCRIPTION,
    defaultHome: Constants.HOME_DIR,
    envVariablePrefix: Constants.ENV_PREFIX,
    webHelpLogoImgPath: path.join(__dirname, "..", "web-help-logo.png"),
    logging: {
        appLogging: {
            logFile: Constants.LOG_LOCATION
        }
    },
    apimlConnLookup: [
        {
            apiId: "ibm.zosmf",
            gatewayUrl: "api/v1",
            connProfType: "zosmf"
        }
    ],
    baseProfile: {
        type: "base",
        schema: {
            type: "object",
            title: "Base Profile",
            description: "Base profile that stores values shared by multiple service profiles",
            properties: {
                host: {
                    type: "string",
                    optionDefinition: Constants.BASE_OPTION_HOST,
                    includeInTemplate: true
                },
                port: {
                    type: "number",
                    optionDefinition: Constants.BASE_OPTION_PORT
                },
                user: {
                    type: "string",
                    secure: true,
                    optionDefinition: Constants.BASE_OPTION_USER,
                    includeInTemplate: true
                },
                password: {
                    type: "string",
                    secure: true,
                    optionDefinition: Constants.BASE_OPTION_PASSWORD,
                    includeInTemplate: true
                },
                rejectUnauthorized: {
                    type: "boolean",
                    optionDefinition: Constants.BASE_OPTION_REJECT_UNAUTHORIZED,
                    includeInTemplate: true
                },
                tokenType: {
                    type: "string",
                    optionDefinition: Constants.BASE_OPTION_TOKEN_TYPE
                },
                tokenValue: {
                    type: "string",
                    secure: true,
                    optionDefinition: Constants.BASE_OPTION_TOKEN_VALUE
                },
                certFile: {
                    type: "string",
                    optionDefinition: Constants.BASE_OPTION_CERT_FILE
                },
                certKeyFile: {
                    type: "string",
                    optionDefinition: Constants.BASE_OPTION_CERT_KEY_FILE
                // },
                // certFilePassphrase: {
                //     type: "string",
                //     secure: true,
                //     optionDefinition: Constants.BASE_OPTION_CERT_FILE_PASSPHRASE
                }
            },
            required: []
        },
        createProfileExamples: [
            {
                options: "base1 --host example.com --port 443 --user admin --password 123456",
                description: "Create a profile called 'base1' to connect to host example.com and port 443"
            },
            {
                options: "base2 --host example.com --user admin --password 123456 --reject-unauthorized false",
                description: "Create a profile called 'base2' to connect to host example.com (default port - 443) " +
                    "and allow self-signed certificates"
            },
            {
                options: "base3 --host example.com --port 1443",
                description: "Create a profile called 'base3' to connect to host example.com and port 1443, " +
                    " not specifying a username or password so they are not stored on disk; these will need to be specified on every command"
            },
            {
                options: "base4 --reject-unauthorized false",
                description: "Create a zosmf profile called 'base4' to connect to default port 443 and allow self-signed certificates, " +
                    "not specifying a username, password, or host so they are not stored on disk; these will need to be specified on every command"
            }
        ],
        updateProfileExamples: [
            {
                options: "base1 --user newuser --password newp4ss",
                description: "Update a base profile named 'base1' with a new username and password"
            }
        ],
        authConfig: [
            {
                serviceName: "apiml",
                handler: __dirname + "/auth/ApimlAuthHandler",
                login: {
                    summary: Constants.APIML_LOGIN_SUMMARY,
                    description: Constants.APIML_LOGIN_DESCRIPTION,
                    examples: [
                        Constants.APIML_LOGIN_EXAMPLE1,
                        Constants.APIML_LOGIN_EXAMPLE2
                    ],
                    options: [
                        Constants.BASE_OPTION_HOST,
                        Constants.BASE_OPTION_PORT,
                        Constants.BASE_OPTION_USER,
                        Constants.BASE_OPTION_PASSWORD,
                        Constants.BASE_OPTION_REJECT_UNAUTHORIZED,
                        Constants.BASE_OPTION_CERT_FILE,
                        Constants.BASE_OPTION_CERT_KEY_FILE
                    ]
                },
                logout: {
                    summary: Constants.APIML_LOGOUT_SUMMARY,
                    description: Constants.APIML_LOGOUT_DESCRIPTION,
                    examples: [
                        Constants.APIML_LOGOUT_EXAMPLE1,
                        Constants.APIML_LOGOUT_EXAMPLE2
                    ],
                    options: [
                        Constants.BASE_OPTION_HOST,
                        Constants.BASE_OPTION_PORT,
                        Constants.APIML_LOGOUT_OPTION_TOKEN_TYPE,
                        Constants.BASE_OPTION_TOKEN_VALUE,
                        Constants.BASE_OPTION_REJECT_UNAUTHORIZED
                    ]
                }
            }
        ]
    },
    authGroupConfig: {
        authGroup: {
            summary: Constants.AUTH_GROUP_SUMMARY,
            description: Constants.AUTH_GROUP_DESCRIPTION
        }
    },
    configAutoInitCommandConfig: {
        handler: __dirname + "/config/auto-init/ApimlAutoInitHandler",
        provider: "APIML",
        autoInit: {
            options: [
                Constants.AUTO_INIT_OPTION_HOST,
                Constants.AUTO_INIT_OPTION_PORT,
                Constants.AUTO_INIT_OPTION_USER,
                Constants.AUTO_INIT_OPTION_PASSWORD,
                Constants.AUTO_INIT_OPTION_REJECT_UNAUTHORIZED,
                Constants.AUTO_INIT_OPTION_TOKEN_TYPE,
                Constants.AUTO_INIT_OPTION_TOKEN_VALUE
            ]
        },
        profileType: "base"
    },
    profiles: [
        {
            type: "zosmf",
            schema: {
                type: "object",
                title: "z/OSMF Profile",
                description: "z/OSMF Profile",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_HOST
                    },
                    port: {
                        type: "number",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PORT,
                        includeInTemplate: true
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_USER
                    },
                    password: {
                        type: "string",
                        secure: true,
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PASSWORD
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_REJECT_UNAUTHORIZED
                    },
                    certFile: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_CERT_FILE
                    },
                    certKeyFile: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_CERT_KEY_FILE
                    },
                    // certFilePassphrase: {
                    //     type: "string",
                    //     secure: true,
                    //     optionDefinition: ZosmfSession.ZOSMF_OPTION_CERT_FILE_PASSPHRASE
                    // },
                    basePath: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_BASE_PATH
                    },
                    protocol: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PROTOCOL
                    },
                    encoding: {
                        type: "string",
                        optionDefinition: {
                            name: "encoding",
                            aliases: ["ec"],
                            description: "The encoding for download and upload of z/OS data set and USS files." +
                                " The default encoding if not specified is IBM-1047.",
                            type: "string"
                        }
                    },
                    responseTimeout: {
                        type: "number",
                        optionDefinition: ZosFilesOptions.responseTimeout
                    }
                },
                required: []
            },
            createProfileExamples: [
                {
                    options: "zos123 --host zos123 --port 1443 --user ibmuser --password myp4ss",
                    description: "Create a zosmf profile called 'zos123' to connect to z/OSMF at host zos123 and port 1443"
                },
                {
                    options: "zos124 --host zos124 --user ibmuser --password myp4ss --reject-unauthorized false",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                        "and allow self-signed certificates"
                },
                {
                    options: "zos125 --host zos125 --port 1443",
                    description: "Create a zosmf profile called 'zos125' to connect to z/OSMF at the host zos125 and port 1443, " +
                        " not specifying a username or password so they are not stored on disk; these will need to be specified on every command"
                },
                {
                    options: "zos126 --reject-unauthorized false",
                    description: "Create a zosmf profile called 'zos126' to connect to z/OSMF on the default port 443 and allow self-signed " +
                        "certificates, not specifying a username, password, or host so they are not stored on disk; these will need to be " +
                        "specified on every command"
                },
                {
                    options: "zosAPIML --host zosAPIML --port 2020 --user ibmuser --password myp4ss --reject-unauthorized false --base-path basePath",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                        "and allow self-signed certificates"
                }
            ],
            updateProfileExamples: [
                {
                    options: "zos123 --user newuser --password newp4ss",
                    description: "Update a zosmf profile named 'zos123' with a new username and password"
                }
            ]
        },
        ZosTsoProfile,
        ZosUssProfile
    ]
};
module.exports = config;
