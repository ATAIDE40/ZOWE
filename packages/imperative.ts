/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

// WIP Imperative version of Zowe CLI
import { IImperativeConfig } from "@brightside/imperative";
import { Constants } from "./Constants";

const config: IImperativeConfig = {
    productDisplayName: Constants.DISPLAY_NAME,
    commandModuleGlobs: ["**/cli/*.definition!(.d).*s"],
    rootCommandDescription: Constants.DESCRIPTION,
    defaultHome: Constants.HOME_DIR,
    envVariablePrefix: Constants.ENV_PREFIX,
    logging: {
        appLogging: {
            logFile: Constants.LOG_LOCATION
        },
    },
    profiles: [
        {
            type: "zosmf",
            schema: {
                type: "object",
                title: "z/OSMF Profile",
                description: "z/OSMF Profile",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: {
                            name: "host",
                            aliases: ["H"],
                            description: "The z/OSMF server host name.",
                            type: "string",
                            required: true,
                        },
                    },
                    port: {
                        type: "number",
                        optionDefinition: {
                            name: "port",
                            aliases: ["P"],
                            description: "The z/OSMF server port.",
                            type: "number",
                            defaultValue: 443,
                        },
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: {
                            name: "user",
                            aliases: ["u"],
                            description: "Mainframe (z/OSMF) user name, which can be the same as your TSO login.",
                            type: "string",
                            implies: ["password"],
                        },
                    },
                    pass: {
                        type: "string",
                        secure: true,
                        optionDefinition: {
                            name: "password",
                            aliases: ["p"],
                            description: "Mainframe (z/OSMF) password, which can be the same as your TSO password.",
                            type: "string",
                            implies: ["user"],
                        },
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: {
                            name: "reject-unauthorized",
                            aliases: ["ru"],
                            description: "Reject self-signed certificates.",
                            type: "boolean",
                            defaultValue: "true",
                        },
                    },
                },
                required: ["host"],
            },
            createProfileExamples: [
                {
                    options: "zos123 --host zos123 --port 1443 --user ibmuser --password myp4ss",
                    description: "Create a zosmf profile called 'zos123' to connect to z/OSMF at host zos123 and port 1443"
                },
                {
                    options: "zos124 --host zos124 --user ibmuser --password myp4ss --reject-unauthorized false",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                    "and allow self-signed certificates"
                }
            ]
        },
        {
            type: "tso",
            schema: {
                type: "object",
                title: "TSO Profile",
                description: "z/OS TSO/E User Profile",
                properties: {
                    account: {
                        type: "string",
                        optionDefinition: {
                            name: "account",
                            aliases: ["a"],
                            description: "Your z/OS TSO/E accounting information.",
                            type: "string",
                            required: true,
                        },
                    },
                    characterSet: {
                        type: "string",
                        optionDefinition: {
                            name: "character-set",
                            aliases: ["cs"],
                            description: "Character set for address space to convert messages and responses from UTF-8 to EBCDIC.",
                            type: "string",
                            defaultValue: "697",
                        },
                    },
                    codePage: {
                        type: "string",
                        optionDefinition: {
                            name: "code-page",
                            aliases: ["cp"],
                            description: "Codepage value for TSO/E address space to convert messages and responses from UTF-8 to EBCDIC.",
                            type: "string",
                            defaultValue: "1047",
                        },
                    },
                    columns: {
                        type: "number",
                        optionDefinition: {
                            name: "columns",
                            aliases: ["cols"],
                            description: "The number of columns on a screen.",
                            type: "number",
                            defaultValue: 80,
                        },
                    },
                    logonProcedure: {
                        type: "string",
                        optionDefinition: {
                            name: "logon-procedure",
                            aliases: ["l"],
                            description: "The logon procedure to use when creating TSO procedures on your behalf.",
                            type: "string",
                            defaultValue: "IZUFPROC",
                        },
                    },
                    regionSize: {
                        type: "number",
                        optionDefinition: {
                            name: "region-size",
                            aliases: ["rs"],
                            description: "Region size for the TSO/E address space.",
                            type: "number",
                            defaultValue: 4096,
                        },
                    },
                    rows: {
                        type: "number",
                        optionDefinition: {
                            name: "rows",
                            description: "The number of rows on a screen.",
                            type: "number",
                            defaultValue: 24,
                        },
                    },
                },
                required: ["account"],
            },
            createProfileExamples: [
                {
                    description: "Create a tso profile called 'myprof' with default settings and JES accounting information of 'IZUACCT'",
                    options: "myprof -a IZUACCT"
                },
                {
                    description: "Create a tso profile called 'largeregion' with a region size of 8192, a logon procedure of MYPROC, and " +
                    "JES accounting information of '1234'",
                    options: "largeregion -a 1234 --rs 8192"
                }
            ]
        },
    ]
};
module.exports = config;
