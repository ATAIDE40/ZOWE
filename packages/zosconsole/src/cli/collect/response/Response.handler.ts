/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import { ICommandHandler, IHandlerParameters, Session } from "@brightside/imperative";
import { CollectCommand, ICollectParms, IConsoleResponse } from "../../../../../zosconsole";

/**
 * Handle to collect a MVS console command response
 * @export
 * @class Handler
 * @implements {ICommandHandler}
 */
export default class Handler implements ICommandHandler {

    public async process(commandParameters: IHandlerParameters) {

        const profile = commandParameters.profiles.get("zosmf");
        const session = new Session({
            type: "basic",
            hostname: profile.host,
            port: profile.port,
            user: profile.user,
            password: profile.pass,
            base64EncodedAuth: profile.auth,
            rejectUnauthorized: profile.rejectUnauthorized,
        });
        const collectParms: ICollectParms = {
            commandResponseKey: commandParameters.arguments.responsekey,
            consoleName: commandParameters.arguments["console-name"],
        };

        const response: IConsoleResponse = await CollectCommand.collect(session, collectParms);

        // Print out the response
        commandParameters.response.console.log(response.commandResponse);

        // Return as an object when using --response-format-json
        commandParameters.response.data.setObj(response);

    }
}
