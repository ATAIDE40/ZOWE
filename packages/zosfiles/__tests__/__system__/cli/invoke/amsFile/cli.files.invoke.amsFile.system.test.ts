/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import * as fs from "fs";
import { Session, TextUtils } from "@brightside/imperative";
import { ITestEnvironment } from "../../../../../../../__tests__/__src__/environment/doc/response/ITestEnvironment";
import { TestEnvironment } from "../../../../../../../__tests__/__src__/environment/TestEnvironment";
import { ITestSystemSchema } from "../../../../../../../__tests__/__src__/properties/ITestSystemSchema";
import { TestProperties } from "../../../../../../../__tests__/__src__/properties/TestProperties";
import { getUniqueDatasetName, runCliScript, stripNewLines } from "../../../../../../../__tests__/__src__/TestUtils";
import { ZosFilesMessages } from "../../../../../src/api/constants/ZosFiles.messages";

let REAL_SESSION: Session;
let testEnvironment: ITestEnvironment;
let systemProps: TestProperties;
let defaultSystem: ITestSystemSchema;
let user: string;
let volume: string;

describe("Invoke AMS CLI", () => {

    // Create the unique test environment
    beforeAll(async () => {
        testEnvironment = await TestEnvironment.setUp({
            tempProfileTypes: ["zosmf"],
            testName: "zos_invoke_ams"
        });

        systemProps = new TestProperties(testEnvironment.systemTestProperties);
        defaultSystem = systemProps.getDefaultSystem();

        REAL_SESSION = new Session({
            user: defaultSystem.zosmf.user,
            password: defaultSystem.zosmf.pass,
            hostname: defaultSystem.zosmf.host,
            port: defaultSystem.zosmf.port,
            type: "basic",
            rejectUnauthorized: defaultSystem.zosmf.rejectUnauthorized,
        });

        user = defaultSystem.zosmf.user.trim().toUpperCase();
        volume = defaultSystem.datasets.list[0].vol;

    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(testEnvironment);
    });

    function createTestAMSStatementFileFromTemplate(templateFile: string, dsname?: string) {
        // replace high-level-qualifier with user value
        const AMSStatement = fs.readFileSync(templateFile).toString();
        const updatedStatement = TextUtils.renderWithMustache(AMSStatement, {DSN: dsname, VOL: volume});
        fs.writeFileSync(templateFile + ".temp", updatedStatement);
        return templateFile + ".temp";
    }

    describe("Success scenarios", () => {

        it("should display invoke and invoke ams help", async () => {
            const response = runCliScript(__dirname + "/__scripts__/invoke_ams_help.sh", testEnvironment);
            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            expect(response.stdout.toString()).toMatchSnapshot();
        });

        it("should invoke ams to create and then delete a VSAM cluster using files containing the appropriate control statement", async () => {
            const dsname = getUniqueDatasetName(defaultSystem.zosmf.user);

            // create a temporary file from the template file that has the proper high level qualifier to create the VSAM file
            let controlStatementFile: string = createTestAMSStatementFileFromTemplate(
                process.cwd() + "/packages/zosfiles/__tests__/__system__/api/methods/invoke/DefineVSAM.ams",
                dsname);

            let response = runCliScript(__dirname + "/__scripts__/command/command_invoke_ams_file.sh",
                testEnvironment, [controlStatementFile]);
            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            let testOutput = stripNewLines(response.stdout.toString());
            expect(testOutput).toContain(ZosFilesMessages.amsCommandExecutedSuccessfully.message);

            // Delete the temp file
            fs.unlinkSync(controlStatementFile);

            // create a temporary file from the template file that has the proper high level qualifier to delete the VSAM file
            controlStatementFile = createTestAMSStatementFileFromTemplate(
                process.cwd() + "/packages/zosfiles/__tests__/__system__/api/methods/invoke/DeleteVSAM.ams",
                dsname);

            response = runCliScript(__dirname + "/__scripts__/command/command_invoke_ams_file.sh",
                testEnvironment, [controlStatementFile]);
            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            testOutput = stripNewLines(response.stdout.toString());
            expect(testOutput).toContain(ZosFilesMessages.amsCommandExecutedSuccessfully.message);

            // Delete the temp file
            fs.unlinkSync(controlStatementFile);
        });
    });

    describe("Expected failures", () => {

        it("should fail due to controlStatements not specified", async () => {
            const response = runCliScript(__dirname + "/__scripts__/command/command_invoke_ams_missing_statement.sh", testEnvironment);
            expect(stripNewLines(response.stderr.toString())).toContain("Syntax Error");
            expect(stripNewLines(response.stderr.toString())).toContain("Missing Positional");
            expect(stripNewLines(response.stderr.toString())).toContain("controlStatements");
        });
    });
});
