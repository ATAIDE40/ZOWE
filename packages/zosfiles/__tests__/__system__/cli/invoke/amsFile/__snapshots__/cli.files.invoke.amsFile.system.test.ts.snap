// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invoke AMS CLI Success scenarios should display invoke and invoke ams help 1`] = `
"================Z/OS FILES INVOKE HELP===============

 DESCRIPTION
 -----------

   Invoke z/OS utilities such as Access Method Services (AMS)

 USAGE
 -----

   bright zos-files invoke [command]

 COMMANDS
 --------

   ams-file | af       Invoke AMS to submit a file    
   ams-statements | as Invoke AMS to submit statements

 GLOBAL OPTIONS
 --------------

   --response-format-json  | --rfj (flag) (optional)

      Produce the command response as a JSON document

   --help  | -h (flag) (optional)

      Display help text

================Z/OS FILES INVOKE HELP WITH RFJ===========
{
  \\"success\\": true,
  \\"message\\": \\"The help was constructed for command: invoke.\\",
  \\"stdout\\": \\"\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   Invoke z/OS utilities such as Access Method Services (AMS)\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   bright zos-files invoke [command]\\\\n\\\\n COMMANDS\\\\n --------\\\\n\\\\n   ams-file | af       Invoke AMS to submit a file    \\\\n   ams-statements | as Invoke AMS to submit statements\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (flag) (optional)\\\\n\\\\n      Produce the command response as a JSON document\\\\n\\\\n   --help  | -h (flag) (optional)\\\\n\\\\n      Display help text\\\\n\\\\n\\",
  \\"stderr\\": \\"\\",
  \\"data\\": \\"\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   Invoke z/OS utilities such as Access Method Services (AMS)\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   bright zos-files invoke [command]\\\\n\\\\n COMMANDS\\\\n --------\\\\n\\\\n   ams-file | af       Invoke AMS to submit a file    \\\\n   ams-statements | as Invoke AMS to submit statements\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (flag) (optional)\\\\n\\\\n      Produce the command response as a JSON document\\\\n\\\\n   --help  | -h (flag) (optional)\\\\n\\\\n      Display help text\\\\n\\\\n\\"
}================Z/OS FILES INVOKE AMS-FILE HELP===========

 COMMAND NAME
 ------------

   ams-file | af

 DESCRIPTION
 -----------

   Submit control statements for execution by Access Method Services (IDCAMS). You
   can use IDCAMS to create VSAM data sets (CSI, ZFS, etc...), delete data sets,
   and more. You must format the control statements exactly as the IDCAMS utility
   expects. For more information about control statements, see the IBM publication
   'z/OS DFSMS Access Method Services Commands'.

 USAGE
 -----

   bright zos-files invoke ams-file <controlStatementsFile> [options]

 POSITIONAL OPTIONS
 ------------------

   controlStatementsFile		 (string)

      The path to a file that contains IDCAMS control statements. Ensure that your
      file does not contain statements that are longer than 255 characters (maximum
      allowed length).

 PROFILE OPTIONS
 ---------------

   --zosmf-profile  | --zosmf-p (string) (optional)

      The name of a (zosmf) profile to load for this command execution.

 GLOBAL OPTIONS
 --------------

   --response-format-json  | --rfj (flag) (optional)

      Produce the command response as a JSON document

   --help  | -h (flag) (optional)

      Display help text


 EXAMPLES
 --------

   Reads the specified file and submits the controls statements:

	 bright zos-files invoke ams-file \\"./path/to/file/MyControlStatements.idcams\\"

================Z/OS FILES INVOKE AMS-FILE HELP WITH RFJ===========
{
  \\"success\\": true,
  \\"message\\": \\"The help was constructed for command: ams-file.\\",
  \\"stdout\\": \\"\\\\n COMMAND NAME\\\\n ------------\\\\n\\\\n   ams-file | af\\\\n\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   Submit control statements for execution by Access Method Services (IDCAMS). You\\\\n   can use IDCAMS to create VSAM data sets (CSI, ZFS, etc...), delete data sets,\\\\n   and more. You must format the control statements exactly as the IDCAMS utility\\\\n   expects. For more information about control statements, see the IBM publication\\\\n   'z/OS DFSMS Access Method Services Commands'.\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   bright zos-files invoke ams-file <controlStatementsFile> [options]\\\\n\\\\n POSITIONAL OPTIONS\\\\n ------------------\\\\n\\\\n   controlStatementsFile\\\\t\\\\t (string)\\\\n\\\\n      The path to a file that contains IDCAMS control statements. Ensure that your\\\\n      file does not contain statements that are longer than 255 characters (maximum\\\\n      allowed length).\\\\n\\\\n PROFILE OPTIONS\\\\n ---------------\\\\n\\\\n   --zosmf-profile  | --zosmf-p (string) (optional)\\\\n\\\\n      The name of a (zosmf) profile to load for this command execution.\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (flag) (optional)\\\\n\\\\n      Produce the command response as a JSON document\\\\n\\\\n   --help  | -h (flag) (optional)\\\\n\\\\n      Display help text\\\\n\\\\n\\\\n EXAMPLES\\\\n --------\\\\n\\\\n   Reads the specified file and submits the controls statements:\\\\n\\\\n\\\\t bright zos-files invoke ams-file \\\\\\"./path/to/file/MyControlStatements.idcams\\\\\\"\\\\n\\\\n\\",
  \\"stderr\\": \\"\\",
  \\"data\\": \\"\\\\n COMMAND NAME\\\\n ------------\\\\n\\\\n   ams-file | af\\\\n\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   Submit control statements for execution by Access Method Services (IDCAMS). You\\\\n   can use IDCAMS to create VSAM data sets (CSI, ZFS, etc...), delete data sets,\\\\n   and more. You must format the control statements exactly as the IDCAMS utility\\\\n   expects. For more information about control statements, see the IBM publication\\\\n   'z/OS DFSMS Access Method Services Commands'.\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   bright zos-files invoke ams-file <controlStatementsFile> [options]\\\\n\\\\n POSITIONAL OPTIONS\\\\n ------------------\\\\n\\\\n   controlStatementsFile\\\\t\\\\t (string)\\\\n\\\\n      The path to a file that contains IDCAMS control statements. Ensure that your\\\\n      file does not contain statements that are longer than 255 characters (maximum\\\\n      allowed length).\\\\n\\\\n PROFILE OPTIONS\\\\n ---------------\\\\n\\\\n   --zosmf-profile  | --zosmf-p (string) (optional)\\\\n\\\\n      The name of a (zosmf) profile to load for this command execution.\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (flag) (optional)\\\\n\\\\n      Produce the command response as a JSON document\\\\n\\\\n   --help  | -h (flag) (optional)\\\\n\\\\n      Display help text\\\\n\\\\n\\\\n EXAMPLES\\\\n --------\\\\n\\\\n   Reads the specified file and submits the controls statements:\\\\n\\\\n\\\\t bright zos-files invoke ams-file \\\\\\"./path/to/file/MyControlStatements.idcams\\\\\\"\\\\n\\\\n\\"
}"
`;
